### ISSUE TITLE
Test: bond dim of MPO tensors depends on the argument "Exact"

### ISSUE THREAD

**Issue by tanlin2013** [chiamin/HybridLeads#9]
When the argument `{"Exact", true}` is not passed to function `toMPO()`, approximated MPO will be computed. While it tends to give equal bond dim on every link, at least for the tight-binding model. 

Suppose we have this hybrid basis,

$$
|k_1\rangle, |k_2\rangle, |k_3\rangle, |k_4\rangle, |i_1\rangle, |i_2\rangle, |i_3\rangle, |i_4\rangle
$$

This means it must right-multiplies the right-singular vectors from $|k_4\rangle$ into $|i_1\rangle$.

In short, the MPO tensors $W$ on each site follows

* $W(k_1) \neq W(k_2) \neq W(k_3) \neq W(k_4)$
* $W(k_4) \neq W(i_1)$
* $W(i_1) \neq W(i_2) = W(i_3) = W(i_4)$

If we don't want this behaviour, the argument `{"Exact", true}` must need to be passed.

**Comment by tanlin2013** [comment]
It's also worthwhile to note that the ordering of MPO indices changes in accordance with the `Exact` argument.

In the absence of `Exact`, the typical ordering follows

$Link \langle In \rangle \rightarrow Link \langle Out \rangle \rightarrow Site \langle In \rangle \rightarrow Site \langle Out \rangle$

in contrast, it follows the ordering

$Site \langle In \rangle \rightarrow Site \langle Out \rangle \rightarrow Link \langle In \rangle \rightarrow Link \langle Out \rangle$

**Comment by tanlin2013** [comment]
Another important aspect is that the unitary basis transformation is applied on the **single-particle basis**, rather than the **many-body basis**. So it requires further checks to see that the transformation won't affect MPO tensors at the real-space side of hybrid basis.

Current observation,
* Without the `Exact` argument $\rightarrow$ MPO tensors are compressed, so we can't directly compare MPO tensors on two bases, as their bond dimensions differ. Also the singular vectors might have been applied into the real-space part of hybrid basis.
* With the `Exact` argument $\rightarrow$ The first 2 MPO tensors (near the contact) in the real-space part of hybrid basis have different bond dimensions to their cousins living in the real-space basis. But the rest of MPO tensors in the real-space part are identical with their cousins living in the real-space basis.

**Comment by chiamin** [comment]
A single-particle basis rotation is no problem, because in many-body basis
this rotation would still be a unitary rotation. I can explain it in our
next meeting. I am almost confident that it would not affect the real-space
part.

Regarding the order, we are not worried about the order of the indices. We
are worried about the order of **states** in each index (in particular the
virtual index). However if you generate an MPO with two sites in real-space
basis, the order of states should be consistent by construction, and then
we use the second-site MPO tensor in iTDVP; then everything should work
consistently.

Tan Tao-Lin ***@***.***> 於 2022年12月8日 週四 晚上11:56寫道：

**Comment by tanlin2013** [comment]
Thank you Chiamin for the reply. Actually I was just meant to leave some notes for the associated commits, so I know what I'm doing and I can backtrack them in the future. 

I need to think more about the unitary transformation, it's not that straightforward to me at this stage. Also my current observation seems against what we expected, so I need to check further and see I didn't do something wrongly.

The order of indices is important to me in the unit test, in the sense that I'm doing element-wise comparison for 2 MPO tensors, so I do need to know the order of indices in prior, so I can check the **states** within by doing the element-wise comparison.

**Comment by chiamin** [comment]
Here's another comment on checking elements in an ITensor. Typically you
need to store and keep track of **all** the indices for all the tensors
you are using, and use these indices to access the correct elements. This
is sometimes annoying but is the design of ITensor.

Tan Tao-Lin ***@***.***> 於 2022年12月9日 週五 上午11:17寫道：

**Comment by tanlin2013** [comment]
I tried to construct the MPO in the analytical manner.

Suppose the contact is in between the *site* $x$ and $x+1$. Then in the hybrid basis, the MPO looks like

\

$$
W[2] = 
\begin{pmatrix}
I_i & -t \cdot U_{xi}^* \cdot a_i^\dagger & -t \cdot U_{xi} \cdot a_i & e_i \cdot n_i \\
0 & \textcolor{red}{-}F_i & 0 & 0 \\
0 & 0 & F_i & 0 \\
0 & 0 & 0 & I_i \\
\end{pmatrix}
$$ 

\

$$
W[i] = 
\begin{pmatrix}
I_i & -t \cdot U_{xi}^* \cdot a_i^\dagger & -t \cdot U_{xi} \cdot a_i & e_i \cdot n_i \\
0 & F_i & 0 & 0 \\
0 & 0 & F_i & 0 \\
0 & 0 & 0 & I_i \\
\end{pmatrix} , \text{for } i \leq x
$$ 

\ 

$$
W[x+1] = 
\begin{pmatrix}
I_i & -t \cdot a_i^\dagger & -t \cdot a_i & -\mu \cdot n_i \\
0 & 0 & 0 & \textcolor{red}{-}a_i \\
0 & 0 & 0 & a_i^\dagger \\
0 & 0 & 0 & I_i \\
\end{pmatrix}
$$

\

$$
W[i] = 
\begin{pmatrix}
I_i & -t \cdot a_i^\dagger & -t \cdot a_i & -\mu \cdot n_i \\
0 & 0 & 0 & a_i \\
0 & 0 & 0 & a_i^\dagger \\
0 & 0 & 0 & I_i \\
\end{pmatrix} , \text{for } i > x
$$

\

where $e_i$'s are the eigenenergies of single-particle Hamiltonian in the subsystem $i \leq x$, and $U_x$ is the x-th eigenvector. 

Note that here $F_i$ is the Jordan-Wigner string which deals with the anti-commutation relation, $a$ and $a^\dagger$ are bosonic creation and annihilation operators, and $W[1]$ is simply the first row of $W[i \leq x]$, and $W[N]$ is the last column of $W[i>x]$. 

Also note that the elements of $W[2]$ and $W[x+1]$ have different sign (in *red*) than other MPO tensors.

**Comment by tanlin2013** [comment]
It appears like this's what AutoMPO generated (without `Exact` argument)

\

$$
W[i] = 
\begin{pmatrix}
I_i & 0 & 0 & 0 \\
e_i \cdot n_i  & I_i & -t \cdot U_{xi}^* \cdot a_i^\dagger & -t \cdot U_{xi} \cdot a_i \\
0 & 0 & F_i & 0 \\
0 & 0 & 0 & F_i \\
\end{pmatrix} , \text{for } i \leq x
$$ 

\

$$
W[i] = 
\begin{pmatrix}
I_i & 0 & 0 & 0 \\
-\mu \cdot n_i & I_i & -t \cdot a_i^\dagger & -t \cdot a_i \\
a_i & 0 & 0 & 0 \\
a_i^\dagger & 0 & 0 & 0 \\
\end{pmatrix} , \text{for } i > x
$$

\

with one minor exception, the coefficient $U_{xi}$ seems to be equally-partitioned into $W[x]$ and $W[x+1]$? This's probably the reason why we observed $W[x+1] \neq W[x+2]$?

According to the order of indices, $\cancel{ Site \langle In \rangle \rightarrow Site \langle Out \rangle \rightarrow Link \langle In \rangle \rightarrow Link \langle Out \rangle }$, maybe we can check further on

* $\cancel{ W[x][:, :, 2, 3] - W[x+1][:, :, 2, 3] = -t \cdot  \left( \sqrt{U_{xi}^{\ast}} - \sqrt{U_{xi}^{\ast}} \right)  \cdot c_i^{\dagger} = 0 }$
* $\cancel{ W[x][:, :, 2, 4] - W[x+1][:, :, 2, 4] = -t \cdot  \left( \sqrt{U_{xi}} - \sqrt{U_{xi}} \right)  \cdot c_i = 0 }$


Update on 2022/12/20 17:20
----------------------------

Turns out I misunderstood the order of indices, it should be the other way around, $Link \langle In \rangle \rightarrow Link \langle Out \rangle \rightarrow Site \langle In \rangle \rightarrow Site \langle Out \rangle$.  

The misleading behaviour is why the previous unit test can still pass? And it turns out `elt()` always returns 0 or infinity when the demanding index is out-of-range, rather than raising an error (https://github.com/ITensor/ITensor/issues/416). So the assertion of $0 == 0$ (or $\infty == \infty$) always passed. Because of this, it's better to check the tag of indices in prior to any element-wise manipulations.

So the reason why $W[x+1] \neq W[x+2]$ is because Jordan-Wigner string is taken into account to manage the anti-commutation relation, and so $W[2]$ and $W[x+1]$ will have different sign in their elements comparing to other MPO tensors.

Update on 2022/12/20 21:20
----------------------------

Also it seems to be sufficient to use `toMPO` without the argument `Exact` (as to answer the original task), even with the presence of long-range coupling. Also it's beneficial as the bond dimension is a constant value.
